File: robot/resources/ExcelReader.py
"""
ExcelReader.py - Librería personalizada para leer datos de prueba desde Excel.
Usada en todos los suites de Robot Framework para el proyecto Biblioteca Pro.
"""
from openpyxl import load_workbook


class ExcelReader:
    """Lee datos de prueba desde hojas de Excel (.xlsx)."""

    ROBOT_LIBRARY_SCOPE = "GLOBAL"

    def __init__(self):
        self._workbook = None
        self._path = None

    def cargar_excel(self, ruta):
        """Carga el archivo Excel en memoria.

        Argumento:
        - ruta: ruta absoluta al archivo .xlsx
        """
        self._workbook = load_workbook(ruta, data_only=True)
        self._path = ruta
        return f"Excel cargado: {ruta}"

    def obtener_datos_hoja(self, nombre_hoja):
        """Devuelve una lista de diccionarios con los datos de la hoja.

        La primera fila se usa como encabezados (claves del diccionario).
        Las filas completamente vacías se omiten.
        """
        if self._workbook is None:
            raise RuntimeError("Primero llama a 'Cargar Excel'.")
        ws = self._workbook[nombre_hoja]
        encabezados = [str(c.value).strip() if c.value else "" for c in ws[1]]
        datos = []
        for fila in ws.iter_rows(min_row=2, values_only=True):
            if all(v is None for v in fila):
                continue
            registro = {encabezados[i]: (str(v).strip() if v is not None else "") for i, v in enumerate(fila)}
            datos.append(registro)
        return datos

    def obtener_fila_por_caso(self, nombre_hoja, id_caso):
        """Devuelve el diccionario de la fila cuyo campo 'id_caso' coincide."""
        filas = self.obtener_datos_hoja(nombre_hoja)
        for f in filas:
            if f.get("id_caso", "").upper() == str(id_caso).upper():
                return f
        raise ValueError(f"Caso '{id_caso}' no encontrado en hoja '{nombre_hoja}'.")

    def obtener_nombres_hojas(self):
        """Devuelve la lista de nombres de hojas del workbook."""
        if self._workbook is None:
            raise RuntimeError("Primero llama a 'Cargar Excel'.")
        return self._workbook.sheetnames

File: java/src/test/java/com/biblioteca/utils/ExcelReader.java
package com.biblioteca.utils;

import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import java.io.*;
import java.util.*;

/**
 * Utilidad para leer datos de prueba desde archivos Excel (.xlsx).
 * Usa Apache POI. Devuelve listas de mapas {columna → valor}.
 */
public class ExcelReader {

    private final Workbook workbook;

    public ExcelReader(String filePath) throws IOException {
        try (FileInputStream fis = new FileInputStream(filePath)) {
            this.workbook = new XSSFWorkbook(fis);
        }
    }

    /**
     * Lee todos los datos de una hoja.
     * La primera fila son encabezados (claves del mapa).
     * Las filas completamente vacías se omiten.
     */
    public List<Map<String, String>> getSheetData(String sheetName) {
        Sheet sheet = workbook.getSheet(sheetName);
        if (sheet == null) {
            throw new IllegalArgumentException("Hoja no encontrada: " + sheetName);
        }

        Row headerRow = sheet.getRow(0);
        List<String> headers = new ArrayList<>();
        for (Cell cell : headerRow) {
            headers.add(cell.getStringCellValue().trim());
        }

        List<Map<String, String>> rows = new ArrayList<>();
        for (int i = 1; i <= sheet.getLastRowNum(); i++) {
            Row row = sheet.getRow(i);
            if (row == null) continue;

            Map<String, String> data = new LinkedHashMap<>();
            boolean hasData = false;
            for (int j = 0; j < headers.size(); j++) {
                Cell cell = row.getCell(j, Row.MissingCellPolicy.CREATE_NULL_AS_BLANK);
                String value = getCellValue(cell);
                data.put(headers.get(j), value);
                if (!value.isEmpty()) hasData = true;
            }
            if (hasData) rows.add(data);
        }
        return rows;
    }

    /**
     * Devuelve la fila cuyo campo "id_caso" coincide (case-insensitive).
     */
    public Map<String, String> getRowByCase(String sheetName, String caseId) {
        return getSheetData(sheetName).stream()
            .filter(row -> caseId.equalsIgnoreCase(row.get("id_caso")))
            .findFirst()
            .orElseThrow(() -> new IllegalArgumentException(
                "Caso '" + caseId + "' no encontrado en hoja '" + sheetName + "'"));
    }

    private String getCellValue(Cell cell) {
        if (cell == null) return "";
        switch (cell.getCellType()) {
            case STRING:  return cell.getStringCellValue().trim();
            case NUMERIC:
                double v = cell.getNumericCellValue();
                return (v == Math.floor(v)) ? String.valueOf((long) v) : String.valueOf(v);
            case BOOLEAN: return String.valueOf(cell.getBooleanCellValue());
            case FORMULA:
                try { return cell.getStringCellValue().trim(); }
                catch (Exception e) { return String.valueOf(cell.getNumericCellValue()); }
            default: return "";
        }
    }

    public void close() throws IOException {
        workbook.close();
    }
}
